### PR TITLE
Combine launch buttons in instance window, persist profiler

### PR DESCRIPTION
**Pull request by TheKodeToad**
https://github.com/PrismLauncher/PrismLauncher/issues/253#issuecomment-1288095352

this is like my third or forth attempt of this or something similar lol
Screencast from 2023-07-06 18-22-11.webm

### CODE REVIEW

**Review by getchoo — DISMISSED** [review]
i'm a big fan of this since i always found it weird how the demo and offline modes were prioritized and there wasn't a similar profiler function as the main window. the only issue i've found is that the launch button dropdown is a bit small and kinda hard to click - especially on windows as seen here
image

i'd recommend making this a little bigger, and probably the same for the main window (maybe the export dropdown too for good measure!)

**Comment by TheKodeToad** [comment]
it's built into qt. i didn't make this. in fact it's a QToolButton, modified to fit into its surroundings which may not always look good on all style implementations! :/

QPushButton's don't support this

**Comment by getchoo** [comment]
yeah that's my bad :/
i thought they could at least be changed with stylesheets but after a bit of messing around i realized that's not the case. that part is good to go, sorry!

**Comment by TheKodeToad** [comment]
In any case, shipping QSS by default doesn't seem like a good idea to me

**Review by getchoo — APPROVED** [review]
almost forgot about this :ioa:

i wish the icons could be bigger, but since we don't really control it and everything else works fine, i think it's ready for a merge :)

File: launcher/minecraft/MinecraftInstance.cpp [hunk 1]
```diff
@@ -239,6 +243,52 @@ QSet<QString> MinecraftInstance::traits() const
...
+        profilers->addAction(profilerAction);
+        profilerAction->setData(profiler.key());
+        profilerAction->setCheckable(true);
+        profilerAction->setChecked(settings()->get("Profiler").toString() == profiler.key());
+
+        QString error;
+        if (profiler.value()->check(&error))
+            profilerAction->setIcon(APPLICATION->getThemedIcon("status-good"));
+        else
+            profilerAction->setIcon(APPLICATION->getThemedIcon("status-bad"));
```
Comment (Scrumplex) [inline launcher/minecraft/MinecraftInstance.cpp]: these icons hide the checkbox. I think we should just disable the items if they are not available

Comment (TheKodeToad) [reply launcher/minecraft/MinecraftInstance.cpp]: problem is, a disabled item can be selected which is weird

Comment (Scrumplex) [reply launcher/minecraft/MinecraftInstance.cpp]: I just tried it locally and it works as expected. I can't click/select the disabled item

Comment (Scrumplex) [reply launcher/minecraft/MinecraftInstance.cpp]:
```suggestion
        profilerAction->setEnabled(profiler.value()->check(&error));
```

Comment (TheKodeToad) [reply launcher/minecraft/MinecraftInstance.cpp]: > I just tried it locally and it works as expected. I can't click/select the disabled item

if the profiler's validity changes

Comment (TheKodeToad) [reply launcher/minecraft/MinecraftInstance.cpp]: kind of a fusion issue. i wonder if there's a workaround.

**Comment by Trial97** [comment]
Will #253 be closed once this PR is merged?

File: launcher/minecraft/MinecraftInstance.cpp (hunk 1 above)
Comment (TheKodeToad) [reply launcher/minecraft/MinecraftInstance.cpp]: Can we just keep it like this because it's Fusion's fault?
Disabling items leads to confusing UX imo

**Comment by TheKodeToad** [comment]
> yeah that's my bad :/ i thought they could at least be changed with stylesheets but after a bit of messing around i realized that's not the case. that part is good to go, sorry!

OK. this should be possible. but maybe done in a separate pr. Depends whether we want our own QStyle :catquake:

File: launcher/minecraft/MinecraftInstance.cpp (hunk 1 above)
Comment (TheKodeToad) [reply launcher/minecraft/MinecraftInstance.cpp]: No response. I'll just resolve and you can unresolve if it's critical

Comment (Scrumplex) [reply launcher/minecraft/MinecraftInstance.cpp]: > Disabling items leads to confusing UX imo

I think it's much more confusing if you can't tell that something is enabled/selected

Comment (TheKodeToad) [reply launcher/minecraft/MinecraftInstance.cpp]: huh? selecting an invalid profiler just makes a dialog box appear on launch

Comment (Scrumplex) [reply launcher/minecraft/MinecraftInstance.cpp]: Screenshot of menu

It is really hard to see that JVisualVM is selected. From a UX view it also doesn't make sense why the `No Profiler` option has a radio button while the two below don't.

Disabling options is just the way to go.

Comment (TheKodeToad) [reply launcher/minecraft/MinecraftInstance.cpp]: image
would you agree this is fine though?
I suppose i can change it just for the sake of fusion :pensive:

Comment (Scrumplex) [reply launcher/minecraft/MinecraftInstance.cpp]: Why not just disable the menu items? That's how it already works to develop. Disabled buttons/menu items is a common UI pattern on desktop systems.

> I suppose i can change it just for the sake of fusion

Please don't forget that Fusion *is* the primary style for Prism Launcher.

**Comment by Scrumplex** [comment]
Make sure to reformat your code to make the pre-commit-check from garnix happy :3

File: launcher/minecraft/MinecraftInstance.cpp (hunk 1 above)
Comment (TheKodeToad) [reply launcher/minecraft/MinecraftInstance.cpp]: I guess disabling is fine. We can't add a tooltip though so it could be confusing.

**Comment by TheKodeToad** [comment]
hmm. why is it failing after i reformatted everything

**Review by Scrumplex — APPROVED** [review]

File: launcher/minecraft/MinecraftInstance.cpp [hunk 2]
```diff
@@ -229,6 +233,49 @@ QSet<QString> MinecraftInstance::traits() const
     return profile->getTraits();
 }
 
+void MinecraftInstance::populateLaunchMenu(QMenu* menu)
```
Comment (Scrumplex) [inline launcher/minecraft/MinecraftInstance.cpp:237]: UI code technically doesn't belong here. But a FIXME should be fine enough :D


```suggestion
// FIXME: move UI code out of MinecraftInstance
void MinecraftInstance::populateLaunchMenu(QMenu* menu)
```

Comment (TheKodeToad) [reply launcher/minecraft/MinecraftInstance.cpp:237]: Where should it be put in future in your opinion?

I thought it was fine as it's not directly working with widgets.

**Pull request merged**